feat(mme): Serialization and Deserialization of ngap state

aniket021997 opened the pull request:
Signed-off-by: aniket021997 <aniket.sonawane@wavelabs.ai>

<!--
    Tag your PR title with the components that it touches along with
    the type of change
    E.g. "fix(orc8r): Fix reindexer race condition" or "feat(agw) ..."
-->

## Summary
Changes made:

- Replaced ngap proto with s1ap proto for feasibility.
- Added changes for serialization and deserialization of supported TA item and list
- Added unit test cases for ngap state serialization and deserialization.

<!-- Enumerate changes you made and why you made them -->

## Test Plan
- Verified basic sanity with ueransim.
- Verified using UT.
- Please find attached logs and pcap for basic registration and pdu session establishment.
mme log:

[mme.log](https://github.com/magma/magma/files/7788577/mme.log)

pcap snap:
![pcap file](https://user-images.githubusercontent.com/95931489/147657381-bdbdcaf5-aa66-45cf-ab8e-fddae934b692.PNG)

validated with unit test cases:
![make_test_oai](https://user-images.githubusercontent.com/95931489/147657417-77a82817-49ac-49b1-9c16-b10f55e82a9c.PNG)


<!--
    How did you test your change? How do you know it works?
    Add supporting screenshots, terminal pastes, etc. as necessary
-->

## Additional Information

- [ ] This change is backwards-breaking

<!--
    If this is a backwards-breaking change, document the upgrade instructions.
    All upgrade instructions for backwards-breaking changes will be aggregated
    in the next release's changelog so this is very important to fill out.
-->

aniket021997 committed:
test(mme): unit test for ngap state serialization and deserialization

Signed-off-by: aniket021997 <aniket.sonawane@wavelabs.ai>

ssanadhya commented on lte/gateway/c/core/oai/test/ngap/test_ngap_state_converter.cpp:
```diff
@@ -78,9 +99,42 @@ TEST_F(NGAPStateConverterTest, NgapStateConversionSuccess) {
+      HASH_TABLE_OK);
+  EXPECT_EQ(
+      hashtable_ts_get(
+          &final_state->amfid2associd, (const hash_key_t) 1,
```
Why is there a lookup for same amf id twice? Shouldn't it be 2 here?

```suggestion
          &final_state->amfid2associd, (const hash_key_t) 2,
```

aniket021997 replied:
Thanks for the review @ssanadhya.
Here we are verifying that whether entries for init state and final state amfid2associd are present. 
On line 122 we are verifying for init state and on line 127 we are verifying for the final state amfid2associd.

ssanadhya replied:
Thanks for the clarification @aniket021997 ! I missed that part. Looks good.

ssanadhya commented on lte/gateway/c/core/oai/tasks/ngap/ngap_state_converter.cpp:
```diff
@@ -24,26 +24,26 @@ using namespace std;
 using namespace magma::lte::oai;
 
-using magma::lte::oai::Ngap_UeDescription;
-using magma::lte::oai::NgapState;
```
What is the motivation for this change? Mixing 4G and 5G terminology will make the code hard to read, e.g. storing gnb_ue_ngap_id as enb_ue_s1ap_id . Given that we are proposing to use the proto object directly in hashtables (https://github.com/magma/magma/issues/11199), this mixed naming will extend over the entire Ngap implementation. Let's keep the proto definitions separate for S1AP and NGAP.

ganeshg87 replied:
@ssanadhya, 
    It was initial feedback from @ulaskozat to reuse existing Proto, designed Stateless feature(#10805) such way to reuse existing Proto.
 cc: @panyogesh

ssanadhya replied:
@ganeshg87 , could you please point me to the initial feedback comment?
Further, which changes in #10805 are you referring to in regards to reusing protos?

ganeshg87 replied:
@ssanadhya, 
    Sorry, PR Id was #10605 (not #10805) design document, amf stateless feature designed to reuse existing 4G infra.mme_nas_state.proto, nas_state.proto reused for amf.

ardzoht replied:
@ganeshg87 
We had an internal discussion with @ssanadhya and @ulaskozat, we agreed it's cleaner to maintain different protobuf structures for 4G and 5G. Main reasoning comes from the fact that we want to separate the naming of 5G and 4G representations once they are eventually extended and used as objects on the internal and redis datastore mappings. The same should be applied to PR #[11302](https://github.com/magma/magma/pull/11032) and #[11005](https://github.com/magma/magma/pull/11005)

aniket021997 commented:
Hi @ssanadhya,
As per the review comment, we replaced the s1ap protos with ngap protos. Also, we have added Ngap_SupportedTaList and Ngap_SupportedTaiItems in Ngap proto file. Please provide your feedback.

- Verified basic sanity with ueransim.
- Validated with UT.
- Please find attached logs and pcap.
[mme.log](https://github.com/magma/magma/files/7926769/mme.log)
![Pcap_snap](https://user-images.githubusercontent.com/95931489/150815056-04c92b7c-8620-4cfd-a314-5aa96fa7f4bf.PNG)
![make_test_oai](https://user-images.githubusercontent.com/95931489/150815103-5a7234ee-1721-4099-b262-90034e44f984.PNG)

ssanadhya reviewed (approved):
LGTM

ssanadhya commented:
@aniket021997 , thanks for revising this PR. Can this be merged now, or do we need other PRs to be merged before this one?

aniket021997 commented:
@ssanadhya thanks for reviewing the PR.
Please merge this PR.
